### PR TITLE
Automatic Error Recovery

### PR DESCRIPTION
**Pull request by jacky-liang**
- expose franka's automatic error recovery function to robot abstract…
- added functions to clear robot state buffers, shared memory buffer, skill lists,
- added functions to create and send RobolibStatus from robolib and corresponding ROS handles to check whether or not robolib is ready. ready is set to false before init finishes and during err recovery.
- added watchdog mechanism through shared memory. robolib resets watch dog to 0, while franka_action_lib increments. when counter gets too high status becomes not ready
- wrapped run_loop's main functionality in while true and try-catch that resets relevant buffers, handles errors, and continue to next iteratino
- removed try-catch from skills as FrankaException is now caught in run_loop
- on franka_action_lib, set skill to be pre-empted if robolib became not ready mid-skill
- on Python side, added appropriate exception raising, handling timeouts, and checking RobolibStatus from ROS
- removed start time from run_loop

### CODE REVIEW

File: iam_robolib/src/run_loop.cpp [hunk 1]
```diff
@@ -501,22 +500,41 @@ void run_loop::setup_current_robot_state_io_thread() {
   });
 }
 
+void run_loop::setup_watchdog_thread() {
+  int io_rate = 100;
```
Comment (mohitsharma0690) [inline iam_robolib/src/run_loop.cpp]: should we do this maybe at a lower rate?

File: iam_robolib/src/run_loop.cpp [hunk 2]
```diff
@@ -556,61 +583,93 @@ void run_loop::run_on_franka() {
...
+      set_robolib_status(false, error_description);
+
+      // Log data
+      std::cout << "Franka exception occurred during control loop. Will reset." << std::endl;
+      std::cerr << error_description << std::endl;
+      robot_state_data_->writeCurrentBufferData();
+      robot_state_data_->printGlobalData(50);
+      logger_.print_error_log();
+      logger_.print_warning_log();
+      logger_.print_info_log();
```
Comment (mohitsharma0690) [inline iam_robolib/src/run_loop.cpp]: I think you can remove these 3 logger statements, we don't use them right now.

File: iam_robolib_common/include/iam_robolib_common/run_loop_process_info.h [hunk 3]
```diff
@@ -152,17 +166,22 @@ class RunLoopProcessInfo {
...
+  int current_memory_region_{1}; 
   int current_sensor_region_{1};
   int current_feedback_region_{1};
-  int current_skill_id_{-1};
+  int current_skill_id_{-1}; 
   int new_skill_id_{-1};
   int done_skill_id_{-1};
   int result_skill_id_{-1};
   char new_skill_description_[1000];
-  size_t new_skill_description_len_=1;
```
Comment (mohitsharma0690) [inline iam_robolib_common/include/iam_robolib_common/run_loop_process_info.h]: I think we should set the length to one, I'm not sure what happens when it goes to 0 and we print it (or get it on python side).

File: src/catkin_ws/src/franka_action_lib/include/franka_action_lib/robolib_status_publisher.h [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+      ros::NodeHandle nh_;
+      ros::Publisher robolib_status_pub_; // NodeHandle instance must be created before this line. Otherwise strange error occurs.
+      std::string topic_name_;
+
+      RobolibStatus last_robolib_status_;
+      bool has_seen_one_robolib_status_;
+      int stale_count = 0;
+
+      double publish_frequency_;
+      franka_action_lib::SharedMemoryHandler shared_memory_handler_;
```
Comment (mohitsharma0690) [inline src/catkin_ws/src/franka_action_lib/include/franka_action_lib/robolib_status_publisher.h]: This is going to invoke the copy constructor, we should use references or pointers and not actual objects.

File: src/catkin_ws/src/franka_action_lib/src/shared_memory_handler.cpp [hunk 5]
```diff
@@ -535,6 +535,37 @@ namespace franka_action_lib
...
+  franka_action_lib::RobolibStatus SharedMemoryHandler::getRobolibStatus()
+  {
+    boost::interprocess::scoped_lock<boost::interprocess::interprocess_mutex> run_loop_info_lock(*run_loop_info_mutex_, boost::interprocess::defer_lock);
+
+    franka_action_lib::RobolibStatus robolib_status;
+    robolib_status.header.stamp = ros::Time::now();
+
+    if (run_loop_info_lock.try_lock()) {
+      // TODO(jacky): 10 roughly equates to 100ms of disconnect from the robolib. This should be placed in a global config file.
+      if (run_loop_process_info_->get_watchdog_counter() > 10) {
```
Comment (mohitsharma0690) [inline src/catkin_ws/src/franka_action_lib/src/shared_memory_handler.cpp]: I think this value should be higher (by default) or some easy way to set it. It seems like a case where for whatever reason the shared memory has some delay and this keeps on resetting the state.

File: iam_robolib/src/run_loop.cpp (hunk 1 above)
Comment (jacky-liang) [reply iam_robolib/src/run_loop.cpp]: sounds good I'll do 50.

**Pull request merged**